contracts/factory/src/contract.rs

```diff
@@ -119,7 +126,123 @@ fn migrate_wallet(
+        new_code_id,
+        msg,
+    }) = tx_msg.clone()
+    {
+        let msg: MigrateMsg = cosmwasm_std::from_slice(&msg)?;
+
+        // Ensure provided msg is proxy msg
+        msg.ensure_is_proxy_msg()?;
+
+        // Ensure migrating the corret wallet at given address
+        if contract_addr != wallet_addr {
+            return Err(ContractError::InvalidMigrationMsg(
+                MigrationMsgError::InvalidWalletAddr,
+            ));
+        }
+        // Ensure user knows the latest supported proxy code id
+        if new_code_id != PROXY_CODE_ID.load(deps.storage)? || new_code_id == wallet_info.code_id {
+            return Err(ContractError::InvalidMigrationMsg(
+                MigrationMsgError::MismatchProxyCodeId,
+            ));
+        };
+    } else {
+        return Err(ContractError::InvalidMigrationMsg(
+            MigrationMsgError::InvalidWasmMsg,
+        ));
+    }
+
+    Ok(Response::new()
+        .add_message(tx_msg)
+        .add_attribute("action", "wallet migration"))
+}
+
+/// Migrates the instantiated `multisig` contract instance to a new code id
+fn migrate_multisig_contract(
+    deps: DepsMut,
+    info: MessageInfo,
+    address: WalletAddr,
+    migration_msg: ProxyMigrationMsg,
+) -> Result<Response, ContractError> {
+    let wallet_addr = match address {
+        WalletAddr::Canonical(canonical_address) => deps.api.addr_humanize(&canonical_address)?,
+        WalletAddr::Addr(human_address) => human_address,
+    };
+
+    let wallet_info: WalletInfo = deps
+        .querier
+        .query_wasm_smart(wallet_addr.clone(), &ProxyQueryMsg::Info {})?;
+
+    // The migration call is either directly called by the user with `ProxyMigrationMsg::DirectMigrationMsg`
+    // or relayed by the proxy relayer via `ProxyMigrationMsg::RelayTx`.
+    //
+    // Different safety checks are applied
+    let tx_msg: CosmosMsg =
+        ensure_is_valid_migration_msg(&deps, info, &wallet_info, &wallet_addr, migration_msg)?;
+
+    // Further checks applied to ensure user has signed the correct relay msg / tx
+    if let CosmosMsg::Wasm(WasmMsg::Migrate {
+        contract_addr,
+        new_code_id,
+        msg,
+    }) = tx_msg.clone()
+    {
+        let msg: MigrateMsg = cosmwasm_std::from_slice(&msg)?;
+
+        // Ensure provided msg is multisig msg
+        msg.ensure_is_multisig_msg()?;
+
+        // Ensure migrating the correct wallet at given address
+        if wallet_info.multisig_address.is_some() {
+            if contract_addr != wallet_addr {
+                return Err(ContractError::InvalidMigrationMsg(
+                    MigrationMsgError::InvalidWalletAddr,
+                ));
+            }
+        } else {
+            return Err(ContractError::InvalidMigrationMsg(
+                MigrationMsgError::MultisigFeatureIsNotSet,
+            ));
+        }
+
+        // Ensure user knows the latest supported proxy code id
+        if new_code_id != PROXY_CODE_ID.load(deps.storage)? {
```

Comment:
Is there any reason we cannot decouple multisig migration and `proxy_code_id` used for the proxy? 
i.e. user may not need to upgrade the proxy contract and can still migrate the multisig?